cmd/snap-confine/mount-support.c

```diff
@@ -505,6 +505,39 @@ static void sc_bootstrap_mount_namespace(const struct sc_mount_config *config)
 	// mount table and software inspecting the mount table may become confused.
 	sc_must_snprintf(src, sizeof src, "%s/proc", SC_HOSTFS_DIR);
 	sc_do_umount(src, UMOUNT_NOFOLLOW | MNT_DETACH);
+	// Detach the both views into /writable: the one from hostfs and the one
+	// directly visible in /writable. Interfaces don't grate access to this
+	// directory and it has a large duplicated view of many mount points.
+	// Note that this is only applicable to ubuntu-core systems.
+	if (config->distro != SC_DISTRO_CLASSIC) {
+		sc_must_snprintf(src, sizeof src, "%s/writable", SC_HOSTFS_DIR);
+		sc_mountinfo *sm SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+		// We use a crude loop to avoid the problem of ordering the mount
+		// points correctly from mostly nested to least nested. This way we
+		// just unmount the first viable candidate we see and start over. This
+		// guarantees that we either didn't want to unmount anything and
+		// finished or we unmounted at least one thing before trying the
+		// sequence again.
+ again:
```

Comment:
I have some mixed feelings about goto for loop like constructs, rather than direct jump to failure/success handling path.

How about something like this: 
```
bool needs_scan = true;
while (needs_scan) {
  sc_mountinfo *sm SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
  sm = sc_parse_mountinfo(NULL);
  if (sm == NULL) {
    die("cannot parse /proc/self/mountinfo");
  }
  
  sc_mountinfo_entry *entry = sc_first_mountinfo_entry(sm);
  bool unmount = false;
  while (entry != NULL) {
    const char *mount_dir = entry->mount_dir;
    ummount = sc_startswith(mount_dir, "/writable") || sc_startswith(mount_dir, src);
    if (umount) {
      sc_do_mount("none", mount_dir, NULL,
                  MS_REC | MS_PRIVATE, NULL);
      sc_do_umount(mount_dir,
                   UMOUNT_NOFOLLOW | MNT_DETACH);
      break;
    }
    entry = sc_next_mountinfo_entry(entry);
  }
  
  /* do another pass if we unmounted anything */
  needs_scan = unmount;
}
```

Reply:
Applied with a small tweak.